refactor: `Monitor`, `ContextDecisionMaking` and `TrialEnv` improvements

gcroci2 opened the pull request:
In this PR, I introduce some changes from #147, which started growing too large. That PR will remain open for a while, so I wanted to avoid diverging too much from dev.

- Added `avg_timesteps_per_trial` property to `TrialEnv`, along with tests.
- Renamed `impl_dim_ring` to `dim_ring`, representing the number of choices in both modalities. This significantly simplifies the overall class, which previously had two different definitions of some private methods depending on implicit/explicit settings.
- Added `plot_training_history` and `evaluate_policy` functionalities to the `Monitor` class, along with docstrings and general improvements.
- Added dependency for recurrent policies (first step toward solving #159).

DaniBodor commented on neurogym/envs/contextdecisionmaking.py:
```diff
@@ -15,21 +15,22 @@ class ContextDecisionMaking(ngym.TrialEnv):
         impl_context_modality: int, which fixed implicit modality to use if `use_expl_context`
             is False (0 or 1).
-        impl_dim_ring: int, number of choices when using implicit representation.
+        dim_ring: int, number of choices in both modalities.
```
I think this statement is more a "historic" thing to explain the change. Not needed going forward, as it is documented above as well.

```suggestion
        dim_ring: int, number of choices.
```

DaniBodor commented on neurogym/envs/contextdecisionmaking.py:
```diff
@@ -81,51 +82,29 @@ def __init__(
             self.timing.update(timing)
 
-        self.abort = False
+        self.abort = abort
```
I see that throughout the package, `self.abort` is always False with no way to change it as a user. Should we change it throughout?
Also, now that it's an outward facing class attribute, maybe it would be better to give it a clearer name. Now, as a user I might get confused and think it immediately aborts the trial or upon any mistake or sth. Maybe something like `strict_fixation`?

DaniBodor commented on neurogym/envs/contextdecisionmaking.py:
```diff
@@ -81,51 +82,29 @@ def __init__(
-            dtype=np.float32,
-            name=obs_space_name,
-        )
+    def _setup_spaces(self):
+        """Setup observation and action spaces for both context types using ring representation."""
```
```suggestion
    def _setup_spaces(self):
        """Setup observation and action spaces using ring representation."""
```

DaniBodor commented on neurogym/envs/contextdecisionmaking.py:
```diff
@@ -135,9 +114,11 @@ def _setup_impl_spaces(self):
         # Action space: fixation + choices
+        action_space_name = {"fixation": 0, "choice": list(range(1, self.dim_ring + 1))}
+
         self.action_space = spaces.Discrete(
-            1 + self.impl_dim_ring,
-            name={"fixation": 0, "choice": list(range(1, self.impl_dim_ring + 1))},
+            n=1 + self.dim_ring,
+            name=action_space_name,
```
given that `self.dim_ring + 1` (or reverse order) come back a few times, would it make sense to give it a separate variable to describes what that represents? Maybe `action_space_size` (if I interpret it correctly).

On the other hand, it is unclear to me why `action_space_name` needs to be defined just to be immediately loaded into `self.action_space` and then not used again

gcroci2 replied:
> given that `self.dim_ring + 1` (or reverse order) come back a few times, would it make sense to give it a separate variable to describes what that represents? Maybe `action_space_size` (if I interpret it correctly).

I think it's okay to leave it as it is, otherwise it's less readable in the code. 

> On the other hand, it is unclear to me why `action_space_name` needs to be defined just to be immediately loaded into `self.action_space` and then not used again

Just for clarity, same for `obs_space_name`.

DaniBodor commented on neurogym/envs/contextdecisionmaking.py:
```diff
@@ -168,55 +149,40 @@ def _new_trial(self, **kwargs: Any) -> dict[str, Any]:  # type: ignore[override]
-            self.add_ob(stim_mod1[i], "stimulus", where=f"stimulus{i + 1}_mod1")
-            self.add_ob(stim_mod2[i], "stimulus", where=f"stimulus{i + 1}_mod2")
+        for i in range(self.dim_ring):
+            self.add_ob(stim_mod1[i], "stimulus", where=f"stim{i + 1}_mod1")
+            self.add_ob(stim_mod2[i], "stimulus", where=f"stim{i + 1}_mod2")
```
why was this abbreviated?

gcroci2 replied:
Because of how it's defined in `obs_space_name`:

```python
        obs_space_name = {
            "fixation": 0,
            **{f"stim{i + 1}_mod1": i + 1 for i in range(self.dim_ring)},
            **{f"stim{i + 1}_mod2": i + 1 + self.dim_ring for i in range(self.dim_ring)},
        }
```

DaniBodor commented on neurogym/wrappers/monitor.py:
```diff
@@ -1,36 +1,53 @@
+    during training and evaluation. It saves relevant behavioral information such as
+    rewards, actions, observations, new trial flags, and ground truth.
+
+    DATA COLLECTION BEHAVIOR:
```
why is this capitalized?

DaniBodor commented on neurogym/wrappers/monitor.py:
```diff
@@ -1,36 +1,53 @@
     """
 
-    metadata: dict[str, str | None] = {  # noqa: RUF012
+    metadata = {  # noqa: RUF012
         "description": (
             "Saves relevant behavioral information: rewards, actions, observations, new trial, ground truth."
         ),
         "paper_link": None,
         "paper_name": None,
     }
```
why is this metadata needed at all? Isn't it basically what the docstring is?

gcroci2 replied:
They're in all envs, it's a class variable which is then used in `TrialEnv` in the `__str__` method

DaniBodor replied:
This isn't an env, though and I don't think that wrapper metadata is used anywhere. 
I do see that other wrappers have the metadata as well. In Gymnasium, usually the envs all have this metadata variable, but the wrappers don't. I guess this is also something that can be a separate refactoring, to remove metadata variables from the wrappers.

DaniBodor commented on neurogym/wrappers/monitor.py:
```diff
@@ -58,11 +75,16 @@ def __init__(
-            self.folder = "tmp"
+            temp_dir = tempfile.mkdtemp(prefix="neurogym_monitor_")
+            self.folder = temp_dir
+        else:
+            self.folder = str(folder)
         Path(self.folder).mkdir(parents=True, exist_ok=True)
-        # seeding
-        self.sv_name = self.folder + self.env.__class__.__name__ + "_bhvr_data_" + name + "_"  # FIXME: use pathlib
+        self.name = name
+
+        # Create save name using pathlib for cross-platform compatibility
+        self.sv_name = Path(self.folder) / f"{self.env.unwrapped.__class__.__name__}_bhvr_data_{name}_"
+
```
I put this FIXME in a while ago, but am actually not sure that is what we want. I actually think it's better to create a normal folder after all, so the user can see it (if using a tempfile location, then I think there's no point in storing it in the first place, is there?)

This now also conflicts with #155 which also changed this and is now merged. Needs to be resolved.

gcroci2 replied:
I solved the conflict by keeping your changes. I agree that a temporary folder doesn't make much sense here.

DaniBodor commented on neurogym/wrappers/monitor.py:
```diff
@@ -87,7 +130,7 @@ def step(self, action):
         if self.sv_stp == "timestep":
             self.t += 1
-        if info["new_trial"]:
+        if info.get("new_trial", False):
```
this looks like it's the opposite of what it used to be. Or am I misinterpreting something?

gcroci2 replied:
This safely checks (now) for the presence of the "new_trial" key. If it exists, it returns the value (True or False). If it doesn't, it returns the default value False. It enters in the if statement only if the key is present and its value is True (same behavior as before)

DaniBodor commented on neurogym/wrappers/monitor.py:
```diff
@@ -115,10 +162,19 @@ def step(self, action):
         for key in self.data:
             self.data[key] = []
 
     def store_data(self, obs, action, rew, info) -> None:
+        """Store data for visualization figures.
+
+        Args:
+            obs: Current observation
+            action: Current action
+            rew: Current reward
+            info: Info dictionary from environment
+        """
```
maybe also add type hints, while you are anyway improving the documentation.

DaniBodor commented on neurogym/wrappers/monitor.py:
```diff
@@ -149,3 +207,228 @@ def store_data(self, obs, action, rew, info) -> None:
             self.gt_mat = []
             self.perf_mat = []
+
+    def plot_training_history(self, window_size=None, figsize=(12, 6), save_fig=True):
```
```suggestion
    def plot_training_history(
        self,
        window_size: int | None = None,
        figsize: tuple[int, int] = (12, 6),
        save_fig: bool = True,
    ):
```

DaniBodor commented on neurogym/wrappers/monitor.py:
```diff
@@ -149,3 +207,228 @@ def store_data(self, obs, action, rew, info) -> None:
+        """Plot training history from saved data files with one data point per trial.
+
+        Args:
+            window_size: Size of the moving average window (default: auto-calculate based on data size)
```
Maybe make explicit that the size is in number of trials, and what the auto-calculated default is.

```suggestion
            window_size: Number of trials for the moving average window (default: ~5% of total trials)
```

gcroci2 replied:
I will remove the window completely

DaniBodor commented on neurogym/wrappers/monitor.py:
```diff
@@ -149,3 +207,228 @@ def store_data(self, obs, action, rew, info) -> None:
             self.perf_mat = []
+
+    def plot_training_history(self, window_size=None, figsize=(12, 6), save_fig=True):
+        """Plot training history from saved data files with one data point per trial.
```
```suggestion
        """Plot rewards and performance history from saved data files with one data point per trial.
```

DaniBodor commented on neurogym/wrappers/monitor.py:
```diff
@@ -149,3 +207,228 @@ def store_data(self, obs, action, rew, info) -> None:
+        return fig
+
+
+def evaluate_performance(env, num_trials=100, model=None, verbose=True):
```
could you add type hints here as well

gcroci2 commented:
I added your suggestions and minor modifications (improvements) @DaniBodor:
- Made `evaluate_performance` a method of Monitor, fix it and rename to `evaluate_policy` (https://github.com/neurogym/neurogym/pull/160/commits/c2d999f93bdb5d9eb7884a7552c9a1441645ae15), and added a test for `evaluate_policy` (https://github.com/neurogym/neurogym/pull/160/commits/b3c21d0aeeeff62a728d6a7eea84cc3622fd1345)
- Simplified `plot_training_history` (https://github.com/neurogym/neurogym/pull/160/commits/c53783e4ae01275b04f01a5b0fdd53d152dce655)
